### ISSUE TITLE
clip : offload to GPU

### ISSUE THREAD

**Issue by ggerganov** [ggerganov/llama.cpp#4061]
With the recent support for running convolutions on the GPU (#4060) we should be able to offload CLIP to run fully on the GPU.

https://github.com/ggerganov/llama.cpp/blob/3d68f364f15778dc326f5024f2e5af1ad6dfddef/examples/llava/clip.cpp#L231-L236

- Implement `ggml_acc` CUDA / Metal kernels
- Avoid `ggml_repeat` where possible using broadcast
- Should use the new `ggml-backend` API (see https://github.com/ggerganov/ggml/blob/master/examples/gpt-2/main-backend.cpp)

**Comment by ggerganov** [comment]
Ideally, CLIP should be supported as a separate model arch in `llama.cpp`, but it will take some extra work to achieve this: https://github.com/abetlen/llama-cpp-python/issues/813#issuecomment-1789221323

We should do it at some point in the future.

**Comment by monatis** [comment]
> Ideally, CLIP should be supported as a separate model arch in llama.cpp, 

Maybe we can start with porting full text and vision encoder parts from my [clip.cpp](https://github.com/monatis/clip.cpp) to `llama.cpp/examples/llava/clip.[h/cpp]` and with the community's testing and feedback support we can polish the implementation gradually. Then we can include it directly in `llama.cpp` as an additional arch once we are confident about its public API and functionality. Or I can continue to develop it externally in that repo and merge later. @ggerganov WDYT?

**Comment by cmp-nct** [comment]
I'd love to see full clip support in llama.cpp soon. 
The current clip is cut down to only support what llava needed, the one of monatis contains a lot more functionality.
Imho we should aim to get the full support of features in, most important is probably llava but as standalone image analysis clip is very valuable.

**Comment by FSSRepo** [comment]
@ggerganov I have implemented broadcast for the ggml_add and ggml_mul operations (only for CPU and CUDA backend). I am just waiting for my pull request to be merged into stable diffusion and will then have some time to incorporate the changes I made in ggml.

**Comment by ggerganov** [comment]
Great! Would be great to PR them in `llama.cpp` (i.e. here) so that we can test CLIP performance.
I think I will be able to help with the Metal implementation

**Comment by FSSRepo** [comment]
> Great! Would be great to PR them in `llama.cpp` (i.e. here) so that we can test CLIP performance. I think I will be able to help with the Metal implementation

See #4205, I think that, for now, we shouldn't merge that pull request until applying the changes I made to ggml in the main project. This way, we'll also have a more comprehensive implementation, eliminating the repeats and all that.

**Comment by y10ab1** [comment]
Do we have any updates on this feature? I am eager to use it!

**Comment by cmp-nct** [comment]
@ggerganov @FSSRepo 
Would be awesome to get this pushed into ggml and llama.cpp
Did you see my discussion on CogLVM ? https://github.com/ggerganov/llama.cpp/discussions/4350
It's a vision model that beats GPT4-vision and should run well on 8-9GB VRAM quantized, it's the first time I have seen anything beating Open-AI. We definitely will need full CLIP offload but the main obstacle is that it has an additional architecture (more than just the 2 layers of projection layer of llava) that connects Vicuna-7 with Big-ViT.
I know it's a bit off topic here, just pushing that because I think it's so significant and totally overlooked

**Comment by FSSRepo** [comment]
@cmp-nct It seems that the architecture of CLIP and Llama differs from the implementation here. The truth is that there will be a lot of work to do if we want to have it here.

**Comment by cmp-nct** [comment]
You are certainly right on the work required, it's likely about as much as the entire clip.cpp has been.

At this point it's the best thing we have in open source for vision, it's right at eye level to GPT4-Vision. 
That's the first time I have seen anything open (or closed) to really compete with the best OpenAI has to offer. 

For "simple vision" llava-1.5 (ShareGPT4V atm) is working great with clip.cpp
If we want real good vision with good OCR then CogVLM would be the current choice.

The only high level alternative is QwenVL which is significantly worse than CogVLM and about the same work to integrate here.

**Comment by ggerganov** [comment]
Done via #4205 and #4696
